### DISCUSSION TITLE
Example of Writing a Locked Register  ...apropos nothing.

### DISCUSSION THREAD

**Discussion by mosqu-ito** [SpenceKonde/DxCore#361] [General]
...and guaranteed to induce vomiting.

<pre>
void PositionFlashWindow(const byte DesiredPosition) {
   constexpr byte MaxWindowPosition =    3;
   constexpr byte FieldMask         = 0xCF;
   constexpr byte FieldPosition     =    4;
   byte T;
   if (DesiredPosition>MaxWindowPosition) return; // Validate requested window position.
   T = NVMCTRL_CTRLB;                             // Take a copy of the whole register.
   T = T & FieldMask;                             // Zero out former contents of field.
   T = T | (DesiredPosition << FieldPosition);    // Replace with requested position.
   CCP = CCP_IOREG_gc;                            // Briefly suspend interrupts & enable write.
   NVMCTRL_CTRLB = T;                             // Promptly write the register.
   }
</pre>

**Comment by SpenceKonde** [comment]
I demand my money back, that only gave me a bruise on my forehead where I slammed it into the desk, it didn't induce vomiting.

I see such terrible code regularly that I've built up a tolerance. I think it may be like arsenic eating (this is a real thing, apparently - people intentionally eat small amounts of arsenic-containing rock for purported benefits...), where exposure to the poison in sub-lethal doses results in a bizarre sort of addiction, and withdrawal symptoms when exposure ceases. I start to get the shakes if I don't see bad code at least once a day (but that rarely happens when you maintain three cores - these cores are chock full of bad code... and the code I write for my own use, well... I've seen inexperienced programmers go blind just looking at it. 

I keep a printout of it near my valuables to foil any would be burglars.... If they manage to make it past my first line of defense - the piles of crap* without being buried under a heap of non-functional electronics, when they get to the bin full of priceless parts (as in, parts that are currently backordered into 2024) they'll still be blinded or driven mad by the unspeakable evil of my code... And anyone who wouldn't be vulnerable to the code because they weren't lesser embedded programmers themselves would have no idea what to steal, and statistically most of what they could grab would be doing me a favor..... 

What's wrong with
```
bool PositionFlashWindow(const byte DesiredPosition) { //return a bool so that failures aren't silent. 
  if (DesiredPosition > 0x03) {
    return true; //indicate error, so you can do `if (PositionFlashWindow(val)) {/* handle error */}`
  }
  uint8_t ctrlb = NVMCTRL.CTRLB; //copying to a temp variable is sort of required here
  // but is always good practice when making changes that require multiple steps to a register, since it's volatile. 
  ctrlb &= ~(NVMCTRL_FLMAP_gm); // zero out flmap bits. 
  ctrlb |= (DesiredPosition << NVMCTRL_FLMAP_gp);
  _PROTECTED_WRITE(NVMCTRL_CTRLB, ctrlb);
  return false;
}
```

`*` "Your room is an appalling junk heap, overflowing with the most disgraceful assortment of deplorable rubbish imaginable!"

**Comment by mosqu-ito** [comment]
You can count on me for such innoculations.  There's nothing at all wrong with your version, indeed I feel quite certain that yours is the example people should follow.  I was kinda _hoping_ you would dot the I's and cross the T's for me.

* You're a bad banana.
